# Exercise 18

## Analyze class homeworks


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
%matplotlib inline
import matplotlib.pyplot as plt

data = pd.read_excel('E18.xlsx')
data

Sexo                                                 T1  \
0     H  Un científico de datos y un ingeniero de datos...   
1     H  Aunque los dos perfiles cuentan con habilidade...   
2     H  Python vs Stata. Analizar información es un pr...   
3     M  Los algoritmos de machine learning buscan pred...   
4     H  Python vs R para análisis de datos. A pesar de...   
5     H  Siendo dos grandes software's de analítica los...   
6     H  Data Scientist y Data Engineer son dos profesi...   
7     M  Las diferencias de un científico de datos y un...   
8     H  Uno de los criterios más importantes en la sel...   
9     H  Diferencia entre un científico de datos y un i...   
10    H  Existen diferencias significativas en el perfi...   
11    H  Por un  lado los ingenieros de datos se centra...   
12    H  Diferencias entre los científicos de datos y l...   
13    H  La diferencia entre un data scientist y un dat...   
14    H  Los algoritmos de Machine Learning tienen como...   
15    M  Machine Learning es un método de análisis de i...   
16    M  Analytics ha tenido una tendencia exponencial ...   
17    H  El duelo entre Data Engineer vs Data Scientist...   
18    H  Los ingenieros de datos tienen como responsabi...   
19    H  Los data scientist y data engineers están embe...   
20    H  La ciencia de datos ha evolucionado mucho a tr...   
21    H  La creación eficiente de modelos no garantiza ...   
22    H  Un ingeniero de datos es la persona que se enc...   
23    H  Para la mayoría de las personas son roles que ...   
24    H  En la actualidad se cuentan con grandes cantid...   
25    H  Python vs R vs SAS for Data Science. Cuando se...   
26    H  Hoy en día, es común en las empresas y grandes...   
27    H                                                NaN   
28    M  Pese a que la ciencia de datos y la ingeniería...   
29    H  En términos de costo, SAS es un software comer...   
30    H  Las diferencias entre un científico de datos y...   
31    H  Dos de las principales razones por las que a p...   
32    H  Entre los programas más populares para realiza...   
33    M  Actualmente para las empresas, los científicos...   
34    H  Un ingeniero de datos es quien se encarga de l...   
35    H  Un ingeniero de datos está encargado de la inf...   
36    H  Diferencia entre un Científico de Datos y un I...   
37    H  El ingeniero de datos se encarga de la arquite...   
38    H  A data Engineer is a support role in the Analy...   
39    M  El  ingeniero de datos y  el científicos de da...   
40    M  Cuando se habla de un Data Engineer, se consid...   
41    H  Aunque un científico de datos y un ingeniero d...   

                                                   T2  \
0                                                 NaN   
1   Aunque el conceso no es total en el tema de ma...   
2   En el marco del Machine Learning encontramos c...   
3   Los algoritmos de machine learning buscan pred...   
4   Tipos de aprendizaje en Machine Learning. El M...   
5                                                 NaN   
6   For Machine Learning there are 3 different typ...   
7   El concepto Machine Learning nace de la indust...   
8   Dentro de los algoritmos usados para programar...   
9   El aprendizaje de máquinas tiene cuatro grande...   
10  Por siglos, la humanidad siempre se ha interes...   
11  1. Aprendizaje supervisado: Los algoritmos de ...   
12  Los tipos de Machine Learning. Existen tres ti...   
13  Tipos de machine learning: Supervised Machine ...   
14                                                NaN   
15                                                NaN   
16  TIPOS DE APRENDIZAJE EN MACHINE LEARNING Como ...   
17  Tipos de aprendizaje de Machine Learning Dentr...   
18  Existen diferentes tipos de técnicas o algorit...   
19                                                NaN   
20  Dentro de los algoritmos de los Algoritmos de ...   
21  De acuerdo al hallazgo que queremos llegar con...   
22                       

# Exercise 18.1

Analyze the writing patterns of each student

In [2]:
data['Sexo'].value_counts()

H    34
M     8
Name: Sexo, dtype: int64

In [3]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from nltk.stem.snowball import SnowballStemmer
from nltk.stem import WordNetLemmatizer
from sklearn import metrics

In [4]:
# check for missing values
data.isnull().sum()

Sexo     0
T1       1
T2      12
T3      12
T4       6
T5       7
T6      13
dtype: int64

In [5]:
data=data.fillna('.')
print(data)

   Sexo                                                 T1  \
0     H  Un científico de datos y un ingeniero de datos...   
1     H  Aunque los dos perfiles cuentan con habilidade...   
2     H  Python vs Stata. Analizar información es un pr...   
3     M  Los algoritmos de machine learning buscan pred...   
4     H  Python vs R para análisis de datos. A pesar de...   
5     H  Siendo dos grandes software's de analítica los...   
6     H  Data Scientist y Data Engineer son dos profesi...   
7     M  Las diferencias de un científico de datos y un...   
8     H  Uno de los criterios más importantes en la sel...   
9     H  Diferencia entre un científico de datos y un i...   
10    H  Existen diferencias significativas en el perfi...   
11    H  Por un  lado los ingenieros de datos se centra...   
12    H  Diferencias entre los científicos de datos y l...   
13    H  La diferencia entre un data scientist y un dat...   
14    H  Los algoritmos de Machine Learning tienen como...   
15    M 

In [6]:
data.isnull().sum()

Sexo    0
T1      0
T2      0
T3      0
T4      0
T5      0
T6      0
dtype: int64

In [7]:
# drop rows with any missing values
data.dropna().shape

(42, 7)

In [8]:
data['concatenada'] = data.apply(lambda x: " ".join([str(x[i]) for i in range(1,6)]),axis=1)

In [9]:
X=data.drop('Sexo',axis=1)
print (X.head())

                                                  T1  \
0  Un científico de datos y un ingeniero de datos...   
1  Aunque los dos perfiles cuentan con habilidade...   
2  Python vs Stata. Analizar información es un pr...   
3  Los algoritmos de machine learning buscan pred...   
4  Python vs R para análisis de datos. A pesar de...   

                                                  T2  \
0                                                  .   
1  Aunque el conceso no es total en el tema de ma...   
2  En el marco del Machine Learning encontramos c...   
3  Los algoritmos de machine learning buscan pred...   
4  Tipos de aprendizaje en Machine Learning. El M...   

                                                  T3  \
0  Los árboles de decisión es uno de los algoritm...   
1                                                  .   
2  Árbol de clasificación Como su nombre lo indic...   
3  Los arboles de decisión pertenecen a los algor...   
4  Tipos de árboles de decisión y aplicaciones

In [10]:
#data.iloc[0][7]

In [11]:
data.dtypes

Sexo           object
T1             object
T2             object
T3             object
T4             object
T5             object
T6             object
concatenada    object
dtype: object

In [12]:
y = data['Sexo'].map({'H':1,'M':0})

In [13]:
cat_val = ["Sexo"]
for col in cat_val:
    print ([col]," : ",data[col].unique())

['Sexo']  :  ['H' 'M']


Genera una serie de pandas con las tareas concatenadas por estudiante para su respectiva validación. Cada fila corresponde a las 6 tareas de cada estudiante.

In [14]:
X_concatenada=pd.Series(data['concatenada'])
print(X_concatenada.head())

0    Un científico de datos y un ingeniero de datos...
1    Aunque los dos perfiles cuentan con habilidade...
2    Python vs Stata. Analizar información es un pr...
3    Los algoritmos de machine learning buscan pred...
4    Python vs R para análisis de datos. A pesar de...
Name: concatenada, dtype: object


Cada frase es separada, de tal forma que los espacios entre ellas seon reemplazados por una coma con el fin de que la frase se convierta en un vector. Se igualan los caracteres, dejan de existir las mayúsculas y se reemplazan por minúsculas.

In [15]:
#X_separada=X_concatenada.str.split(expand=False)
X_separada=X_concatenada.str.lower()
X_separada=X_separada.str.split(expand=False)
print(X_separada.head())


0    [un, científico, de, datos, y, un, ingeniero, ...
1    [aunque, los, dos, perfiles, cuentan, con, hab...
2    [python, vs, stata., analizar, información, es...
3    [los, algoritmos, de, machine, learning, busca...
4    [python, vs, r, para, análisis, de, datos., a,...
Name: concatenada, dtype: object


Se crea una variable tipo counter de la información de las tareas por alumno, donde se cuenta la frecuencia de aparición de las palabras en los textos de cada estudiante.

In [43]:
from collections import Counter
info_alumno = Counter()
for alumno in range(0,len(X_separada)):
    info_alumno[alumno]=Counter(" ".join(X_separada[alumno]).split(" ")).items()
    #Counter(" ".join(X_separada[alumno]).split(" ")).items()

Estudiante 40. De acuerdo a las tareas entregadas este usuario corresponde a mis tareas, a pesar de que aparace que no entregué la última tarea, pero yo si se que la hice ;).

In [17]:
print(info_alumno[40])

dict_items([('cuando', 7), ('se', 44), ('habla', 3), ('de', 151), ('un', 27), ('data', 2), ('engineer,', 1), ('considera', 2), ('que', 55), ('son', 10), ('personas', 1), ('interactúan', 1), ('y', 39), ('manejan', 1), ('los', 39), ('datos', 15), ('desde', 3), ('las', 20), ('fuentes', 2), ('primarias', 1), ('información,', 2), ('contemplan', 1), ('la', 46), ('falta', 1), ('información', 7), ('real', 1), ('dentro', 6), ('una', 20), ('organización;', 1), ('el', 34), ('desarrollo', 2), ('sus', 1), ('competencias', 2), ('está', 5), ('enfocado', 1), ('hacia', 4), ('lo', 8), ('computacional,', 1), ('programación,', 1), ('almacenamiento', 1), ('eficiencia', 1), ('eficacia', 1), ('en', 42), ('su', 9), ('procesamiento.', 1), ('punto', 1), ('vista', 1), ('más', 11), ('estricto,', 1), ('científico', 3), ('enfoca', 2), ('conocimiento', 4), ('manejo,', 1), ('uso,', 1), ('tratamiento', 1), ('dichos', 1), ('posteriormente', 1), ('transformarán', 1), ('input', 1), ('para', 21), ('diseño', 1), ('modelami

Las palabras más frecuentemente utilizadas corresponden a preposiciones, conectores y algunos adverbios.

Seleccionando otro estudiante, de género masculino.

In [45]:
print(info_alumno[20])

dict_items([('la', 28), ('ciencia', 1), ('de', 92), ('datos', 2), ('ha', 1), ('evolucionado', 1), ('mucho', 3), ('a', 13), ('través', 1), ('las', 10), ('últimas', 1), ('décadas', 1), ('obteniendo', 1), ('como', 10), ('uno', 2), ('sus', 3), ('desarrollos', 2), ('más', 6), ('significativos', 1), ('los', 20), ('algoritmos', 11), ('machine', 16), ('learning.', 2), ('sin', 2), ('embargo,', 1), ('aunque', 2), ('learning', 11), ('proporcionan', 1), ('un', 13), ('gran', 3), ('avance', 1), ('para', 10), ('el', 26), ('análisis', 1), ('predictivo', 2), ('conjunto', 3), ('información', 8), ('presentan', 1), ('algunas', 2), ('deficiencias', 1), ('importantes', 2), ('que', 20), ('limitan', 1), ('precisión', 3), ('resultados', 4), ('obtenidos:', 1), ('construcción', 3), ('del', 9), ('algoritmo', 10), ('consiste', 1), ('en', 18), ('entrenamiento', 2), ('sistema', 2), ('sea', 2), ('capaz', 1), ('comprender', 1), ('comportamiento', 1), ('información,', 2), ('por', 13), ('lo', 4), ('necesita', 2), ('gran

Es más conciso que yo en las palabras, no están tan marcado el uso de preposiciones ni conectores como con mis párrafos. Particularmente llama la atención la aparación de la palabra de inicialmente 151, y para el caso del alumno 20, solo 92 veces.

In [18]:
#reduce(lambda x: x['X_separada'], X_separada[0])

In [19]:
# Term Frequency
vect = CountVectorizer()
dftareas = pd.DataFrame(vect.fit_transform(data['concatenada']).toarray(), columns=vect.get_feature_names())
dftareas

000  10  100  101  102  11  112  12  121  123   ...    última  últimas  \
0     0   0    0    1    0   0    0   0    1    1   ...         1        0   
1     0   0    0    0    0   0    0   0    1    0   ...         0        0   
2     0   0    0    0    0   0    0   0    1    0   ...         0        0   
3     0   0    0    0    0   0    0   0    0    0   ...         0        0   
4     0   0    0    0    0   0    0   0    0    0   ...         0        0   
5     0   0    0    0    0   0    0   0    0    0   ...         0        0   
6     0   0    0    0    0   0    0   0    1    0   ...         0        0   
7     0   0    0    0    0   0    0   0    1    0   ...         0        0   
8     0   0    0    0    0   0    0   2    0    0   ...         0        0   
9     0   0    0    0    0   0    0   0    2    0   ...         0        0   
10    0   0    0    0    0   0    0   0    0    0   ...         0        0   
11    0   0    0    0    0   0    0   0    0    0   ...         0        0   
12    0   1    0    0    0   0    0   0    1    0   ...         1        0   
13    0   0    0    0    0   0    0   0    1    0   ...         0        0   
14    0   0    0    0    0   0    0   0    0    0   ...         0        0   
15    0   0    0    0    0   0    1   0    0    0   ...         0        0   
16    0   0    0    0    1   0    0   0    1    0   ...         1        0   
17    0   0    0    0    0   0    0   0    0    0   ...         0        0   
18    0   0    0    0    0   0    0   0    0    0   ...         1        0   
19    0   0    0    1    0   0    0   0    0    1   ...         0        1   
20    0   0    0    0    0   0    0   0    0    0   ...         0        1   
21    0   2    1    0    1   0    0   0    6    0   ...         0        0   
22    0   0    0    0    0   0    0   0    0    0   ...         0        0   
23    0   0    1    0    0   0    0   0    0    0   ...         1        0   
24    0   0    0    0    1   0    0   0    2    0   ...         0        0   
25    0   0    0    0    0   0    0   0    0    0   ...         0        0   
26    0   1    0    0    0   0    0   0    1    0   ...         0        0   
27    0   0    0    0    0   0    0   0    1    0   ...         0        0   
28    0   0    0    0    0   0    0   0    1    0   ...         0        0   
29    0   1    1    0    0   0    0   0    1    0   ...         0        0   
30    1   1    0    0    0   0    0   0    0    0   ...         0        0   
31    0   0    0    0    0   0    0   0    0    0   ...         0        0   
32    0   0    0    0    0   0    0   0    1    0   ...         0        0   
33    0   0    1    0    0   0    0   0    0    0   ...         0        0   
34    0   0    0    0    0   0    0   0    1    0   ...         0        0   
35    0   0    0    0    0   0    0   0    1    0   ...         0        0   
36    0   1    0    0    0   1    1   1    0    0   ...         0        0   
37    0   0    0    0    1   0    0   0    2    0   ...         0        0   
38    0   0    0    0    0   0    0   0    0    0   ...         0        0   
39    0   0    0    0    0   0    0   0    1    0   ...         0        0   
40    0   0    0    0    0   0    0   0    0    0   ...         0        0   
41    0   0    0    0    0   0    2   0    0    0   ...         1        0   

    último  últimos  única  únicamente  único  únicos  útil  útiles  
0        3        0      0           0      0       0     0       0  
1        0        0      0           0      1       0     1       0  
2        0        0      0           0      0       0     0       0  
3        1        1      0           0      0       0     1       0  
4        0        0      0           0      0       0     0       0  
5        0        0      1           0      2       0     2       0  
6        0        0      0           0      0       0     0       0  
7        0        0      0           0      0       0     0       0  
8        0        0

In [20]:
max(dftareas)

'útiles'

In [21]:
max(dftareas.iloc [0])

144

In [22]:
sorted(vect.vocabulary_)

['000',
 '10',
 '100',
 '101',
 '102',
 '11',
 '112',
 '12',
 '121',
 '123',
 '130000',
 '14',
 '140',
 '140k',
 '151',
 '151k',
 '15872',
 '16',
 '165',
 '17',
 '170',
 '170k',
 '172',
 '179',
 '180',
 '183',
 '183k',
 '1980',
 '1991',
 '1996',
 '20',
 '200',
 '2000',
 '2007',
 '2009',
 '2010',
 '2013',
 '2014',
 '2016',
 '2017',
 '21',
 '21659',
 '24',
 '25',
 '262',
 '30',
 '300',
 '32',
 '33',
 '400',
 '50',
 '500',
 '57',
 '65',
 '659',
 '76',
 '81',
 '82',
 '83k',
 '84',
 '86',
 '90',
 '92',
 '93',
 '94',
 '95',
 'abajo',
 'abarcan',
 'abarcar',
 'abarque',
 'abiertas',
 'abierto',
 'abiertos',
 'ability',
 'aboles',
 'abordan',
 'abordar',
 'about',
 'above',
 'abren',
 'abreviatura',
 'abstraer',
 'academia',
 'academias',
 'académica',
 'académicas',
 'acceden',
 'acceder',
 'accepts',
 'accesible',
 'accesibles',
 'acceso',
 'accesos',
 'acciones',
 'accionistas',
 'acción',
 'according',
 'accuracy',
 'accurately',
 'aceptar',
 'acerca',
 'acercan',
 'acertada',
 'acertadas'

# Exercise 18.2

Evaluate the similarities of the homeworks of the students

tip: https://github.com/orsinium/textdistance

In [23]:
# built-in
import json
import os
import os.path
from collections import defaultdict
from copy import deepcopy
from importlib import import_module

In [24]:
class LibrariesManager(object):
    def __init__(self):
        self.libs = defaultdict(list)

    def register(self, alg, lib):
        """Register new lib
        """
        self.libs[alg].append(lib)

    def optimize(self):
        """Sort algorithm implementations by speed.
        """
        # load benchmarks results
        with open(LIBRARIES_FILE, 'r') as f:
            libs_data = json.load(f)
        # optimize
        for alg, libs_names in libs_data.items():
            libs = self.get_libs(alg)
            if not libs:
                continue
            # drop slow libs
            self.libs[alg] = [lib for lib in libs if [lib.module_name, lib.func_name] in libs_names]
            # sort libs by speed
            self.libs[alg].sort(key=lambda lib: libs_names.index([lib.module_name, lib.func_name]))

    def get_algorithms(self):
        """Get list of available algorithms.
        """
        return list(self.libs.keys())

    def get_libs(self, alg):
        """Get libs list for algorithm
        """
        if alg not in self.libs:
            return []
        return self.libs[alg]

    def clone(self):
        """Clone library manager prototype
        """
        obj = self.__class__()
        obj.libs = deepcopy(self.libs)
        return obj

In [25]:
class LibraryBase(object):
    func = NotImplemented

    def __init__(self, module_name, func_name, attr=None, presets=None, conditions=None):
        self.module_name = module_name
        self.func_name = func_name
        self.presets = presets
        self.conditions = conditions
        self.attr = attr

    def check_conditions(self, obj, *sequences):
        # external libs can compare only 2 strings
        if len(sequences) != 2:
            return False
        if not self.conditions:
            return True
        for name, value in self.conditions.items():
            if getattr(obj, name) != value:
                return False

        return True

    def prepare(self, *sequences):
        return sequences

    def get_function(self):
        if self.func is NotImplemented:
            # import module
            try:
                module = import_module(self.module_name)
            except ImportError:
                self.func = None
                return

            # get object from module
            obj = getattr(module, self.func_name)
            # init class
            if self.presets is not None:
                obj = obj(**self.presets)
            # get needed attribute
            if self.attr:
                obj = getattr(obj, self.attr)
            self.func = obj

        return self.func

In [26]:
class TextLibrary(LibraryBase):
    def check_conditions(self, obj, *sequences):
        if not super(TextLibrary, self).check_conditions(obj, *sequences):
            return False
        # compare only by letters
        if getattr(obj, 'qval', 0) != 1:
            return False
        return True

    def prepare(self, *sequences):
        # convert list of letters to string
        if isinstance(sequences[0], (tuple, list)):
            sequences = list(map(lambda x: u''.join(x), sequences))

        # convert to unicode for python2
        try:
            sequences = list(map(unicode, sequences))
        except NameError:
            pass

        return sequences


In [27]:
class SameLengthLibrary(LibraryBase):
    def check_conditions(self, obj, *sequences):
        # if not super(SameLengthLibrary, self).check_conditions(obj, *sequences):
           # return False
        # compare only same length iterators
        if min(map(len, sequences)) != max(map(len, sequences)):
            return False
        return True


In [28]:
# class SameLengthTextLibrary(SameLengthLibrary, TextLibrary):
   # pass

In [29]:
from difflib import SequenceMatcher as SM
print(SM(None,data.concatenada[2], data.concatenada[3]).ratio())

0.016386510201924092


In [30]:
def similarity(data):
    from difflib import SequenceMatcher as SM
    results=pd.DataFrame(0, index=data.index, columns=data.index)
    for i in results.index:    
        for j in results.index:
            if pd.isna(data.concatenada[i])==1 or pd.isna(data.concatenada[j])==1:
                results.loc[j,i]=0            
            else:
                results.loc[j,i]=SM(None,data.concatenada[i],data.concatenada[j]).ratio()
                
                    
            globals()['Tx%s' % i] = data.loc[i, 'concatenada']
            #results.loc[j,i]=SM(None,data.Text[i],data.Text[j]).ratio()
            #print(SM(None,data.Text[i],data.Text[j]).ratio())
    
    #assert results.shape == (data.shape[0], data.shape[0])
    return results

In [31]:
similarity(data)

0         1         2         3         4         5         6   \
0   1.000000  0.018796  0.013587  0.021479  0.012242  0.016023  0.025651   
1   0.012781  1.000000  0.014437  0.017634  0.022048  0.011556  0.017125   
2   0.012293  0.012694  1.000000  0.017972  0.029024  0.021882  0.021931   
3   0.015099  0.009429  0.016387  1.000000  0.009242  0.015904  0.011086   
4   0.008087  0.019569  0.021433  0.012433  1.000000  0.004952  0.008514   
5   0.026163  0.021630  0.029715  0.019085  0.026326  1.000000  0.021552   
6   0.027847  0.016586  0.017109  0.014253  0.007435  0.020609  1.000000   
7   0.041767  0.020921  0.019920  0.054765  0.015739  0.014551  0.015444   
8   0.022968  0.031471  0.022454  0.016757  0.017186  0.031721  0.023485   
9   0.032519  0.023165  0.028331  0.024621  0.015781  0.016322  0.022983   
10  0.020109  0.014728  0.013788  0.013811  0.017098  0.019662  0.002897   
11  0.018333  0.013650  0.019884  0.014151  0.019729  0.013329  0.012180   
12  0.028678  0.024055  0.023618  0.017837  0.014840  0.017869  0.017197   
13  0.016874  0.037250  0.022026  0.017631  0.013529  0.019324  0.024625   
14  0.008798  0.008778  0.017447  0.029311  0.007108  0.019087  0.004827   
15  0.010987  0.009549  0.010799  0.009480  0.004050  0.012994  0.017975   
16  0.021683  0.048426  0.015208  0.018660  0.026910  0.019478  0.015374   
17  0.015134  0.013588  0.016853  0.002127  0.021697  0.021479  0.015756   
18  0.019622  0.011130  0.013118  0.021922  0.008955  0.035480  0.022536   
19  0.016889  0.016284  0.019384  0.013205  0.013806  0.015378  0.017230   
20  0.009357  0.010119  0.009015  0.022811  0.003452  0.012756  0.005968   
21  0.023003  0.034070  0.017320  0.017555  0.011217  0.014009  0.013324   
22  0.012289  0.009871  0.009390  0.013092  0.007063  0.014556  0.015005   
23  0.009137  0.010961  0.021134  0.012722  0.012119  0.020269  0.010236   
24  0.026360  0.021046  0.007990  0.015848  0.021636  0.015182  0.008781   
25  0.019120  0.018470  0.026720  0.018438  0.022421  0.017108  0.015255   
26  0.021658  0.011858  0.031631  0.016459  0.013312  0.018464  0.019548   
27  0.015760  0.026931  0.011834  0.016674  0.012715  0.021517  0.013884   
28  0.025544  0.008909  0.011585  0.013359  0.011017  0.020995  0.005886   
29  0.015058  0.023492  0.012669  0.016526  0.015467  0.063029  0.005038   
30  0.040516  0.007441  0.026020  0.015107  0.011587  0.020930  0.015428   
31  0.012161  0.017689  0.025480  0.014251  0.016021  0.017952  0.011498   
32  0.033204  0.018209  0.035613  0.013664  0.033688  0.022653  0.028264   
33  0.027899  0.024972  0.014513  0.013463  0.021496  0.027816  0.012784   
34  0.030284  0.017631  0.013499  0.007786  0.014686  0.020850  0.012207   
35  0.029541  0.023643  0.041743  0.017400  0.018650  0.022303  0.024581   
36  0.038278  0.013417  0.019494  0.010388  0.007991  0.011578  0.025493   
37  0.020586  0.025047  0.035015  0.030112  0.007432  0.037270  0.026963   
38  0.003190  0.006601  0.002043  0.000907  0.004787  0.003904  0.006926   
39  0.024451  0.017569  0.019969  0.019324  0.004925  0.022875  0.018007   
40  0.018502  0.010052  0.012619  0.019722  0.005901  0.019836  0.006757   
41  0.020562  0.028454  0.011391  0.014749  0.006873  0.015125  0.016196   

          7         8         9     ...           32        33        34  \
0   0.037126  0.019930  0.028388    ...     0.029337  0.022709  0.011591   
1   0.020203  0.024319  0.018776    ...     0.028432  0.016856  0.021229   
2   0.011771  0.020040  0.011269    ...     0.036010  0.005698  0.018691   
3   0.014765  0.014261  0.021608    ...     0.008459  0.009647  0.015059   
4   0.010399  0.014279  0.010301    ...     0.032296  0.006270  0.017818   
5   0.009837  0.036579  0.017296    ...     0.022812  0.017089  0.011024   
6   0.006043  0.021267  0.019153    ...     0.023794  0.011632  0.019309   
7   1.000000  0.019501  0.038890    ...     0.018555  0.027591  0.023267   
8   0.016601  1.000000  0.019289    ...     0.049470  0.0

Teniendo en cuenta la comparación realizada en el punto anterior, se procede a revisar la similaridad entre los párrafos escritos por el estudiante 20 y el estudiante 40, 

# Exercise 18.3

Create a classifier to predict the sex of each student



In [32]:
X=data['concatenada']

In [33]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/lucilaevelysnoriegafornaris/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [34]:
def split_into_lemmas(text):
    text = text.lower()
    words = text.split()
    #print(words)
    return [wordnet_lemmatizer.lemmatize(word) for word in words]

In [35]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)
vect = TfidfVectorizer(max_df=0.4,ngram_range=(1, 2),analyzer=split_into_lemmas)
X_train_vec = vect.fit_transform(X_train)

In [36]:
X_test_vec=vect.transform(X_test)

In [37]:
rf=RandomForestClassifier(n_estimators=10)
rf.fit(X_train_vec, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [38]:
y_pred=rf.predict(X_test_vec)

In [39]:
cols = ['p_G']

In [40]:
#print(y_pred)
res = pd.DataFrame(y_pred, columns=cols)
print(res)

    p_G
0     1
1     1
2     1
3     1
4     1
5     1
6     1
7     1
8     1
9     1
10    1
11    1
12    1
13    1
14    1
15    1
16    1
17    1
18    1
19    1
20    1


In [41]:
from sklearn.metrics import r2_score, roc_auc_score
roc_auc_score(y_test, y_pred, average='macro')

0.5

Como son tan pocos datos es muy complejo lograr una precisión alta respecto al sexo de quine escribió cada texto.